services/api/src/utils/schema.js

```diff
@@ -85,6 +85,20 @@ function createSchema(attributes = {}, options = {}) {
     return this.save();
   };
 
+  schema.post(/^init|save/, function () {
+    if (!this.$locals.auditOriginal) {
```

Comment:
so I don't want to beat a dead horse but remind me again what we lose if we don't do this? It just seems like a lot of hacking we're doing here on every save on all models just for the sake of the audit trail... If the entire facility doesn't work without it then maybe I get it but if for example it's just for the sake of making populated objects work slightly better, then I don't think it's worth it.

I'm not really expecting diffs on populated objects anyway I think as populated by definition means they're different models, I'm only interested in what the ref changed to. Remember, as long as we have the refs we could potentially fetch them manually if needed. I feel like this is perfectly fine.

Reply:
All I am doing here is store the original object how it was before it was changed + storing the directModifiedPaths.

Another point I want to clarify, the tracking of populated objects are outside the scope. Nothing is tracking besides the reference (objectId). E.g. Consider the `product.shop` attribute I want to track when the product.shop objectId is changed. 

The reason for this:
First this avoid us from splitting AuditLog.getObjectFields from the AuditLog.append
Before this is what you have write
```
const { user } = ctx.state 
user.assign(ctx.request.body);
const objectFields = AuditLog.getObjectFields(user, ["email"]);
await user.save() 
await AuditLog.append("updated user", ctx, ...objectFields);
```
Now you can write, as the original we still have access to via $locals and we still have access to the what fields has changed.
```
const { user } = ctx.state 
user.assign(ctx.request.body);
await user.save() 
await AuditLog.append("updated user", ctx, ...AuditLog.getObjectFields(user, ["email"]));
```

Think thats worth it for that. I hoped that mongoose would store the original object, but it does not.

The second thing that comes for free is that it also allows me fix handle the objectId issue (its always reported as modified) because i have the original state of the object.

Reply:
ok I'm starting to understand so we lose the ability to diff essentially, nothing to do with populating... makes sense I think but a few comments I'll add inline